src/olympia/blocklist/utils.py

```diff
@@ -153,24 +151,77 @@ def save_guids_to_blocks(guids, submission, *, fields_to_set):
     for block in blocks:
         change = bool(block.id)
         if change:
-            block_obj_before_change = Block(
-                min_version=block.min_version, max_version=block.max_version
-            )
             setattr(block, 'modified', modified_datetime)
-        else:
-            block_obj_before_change = None
         for field, val in common_args.items():
             setattr(block, field, val)
         block.average_daily_users_snapshot = block.current_adu
         block.save()
+        # And now update the BlockVersion instances - instances to add first
+        block_versions_to_create = []
+        for version in block.addon_versions:
+            if version.id in submission.changed_version_ids and (
+                not change or not version.is_blocked
+            ):
+                block_version = BlockVersion(block=block, version=version)
+                block_versions_to_create.append(block_version)
+                version.blockversion = block_version
+        BlockVersion.objects.bulk_create(block_versions_to_create)
+
         if submission.id:
             block.submission.add(submission)
         block_activity_log_save(
             block,
             change=change,
             submission_obj=submission if submission.id else None,
-            old_obj=block_obj_before_change,
         )
-        disable_addon_for_block(block)
+        disable_versions_for_block(block, submission)
+        if submission.disable_addon:
+            if block.addon.status == amo.STATUS_DELETED:
+                try:
+                    block.addon.deny_resubmission()
+                except GuidAlreadyDeniedError:
+                    pass
+            else:
+                block.addon.update(status=amo.STATUS_DISABLED)
+
+    return blocks
+
+
+def delete_versions_from_blocks(guids, submission, *, fields_to_set):
```

Comment:
Is this logic covered by any tests ? or is it part of the things you were considering in a follow-up ?

Reply:
TestBlockAdminDelete tests this, by deleting one block completely (all blocked versions), and just removing a single blocked version from the other.